cmd/check_run.go

```diff
@@ -0,0 +1,75 @@
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/engine"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/formatters"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/runtime"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var checkRunCmd = &cobra.Command{
+	Use:   "run",
+	Short: "Run a single check inside of podman unshare",
+	Long: `This command will run a check while wrapped inside of podman unshare.
+It is an internal command, and is not meant to be called by the user.
+It takes its input from environment variables only.`,
+	Hidden: true,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// Expect an environment variable named PREFLIGHT_CHECK_EXEC
+		// It will specify the one check to execute inside of the podman unshare
+		// environment.
+		enabledCheck, ok := os.LookupEnv("PREFLIGHT_EXEC_CHECK")
+		if !ok {
+			log.Error("Enabled check envvar not specified")
+			return errors.New("required environment variable PREFLIGHT_EXEC_CHECK not specified")
+		}
+		image, ok := os.LookupEnv("PREFLIGHT_EXEC_IMAGE")
+		if !ok {
+			log.Error("Operator image envvar not specified")
+			return errors.New("required environment variable PREFLIGHT_EXEC_IMAGE not specified")
+		}
+
+		cfg := runtime.Config{
+			Image:          image,
+			EnabledChecks:  []string{enabledCheck},
+			ResponseFormat: DefaultOutputFormat,
+			Mounted:        true,
+		}
+
+		engine, err := engine.NewForConfig(cfg)
+		if err != nil {
+			return err
+		}
+
+		formatter, err := formatters.NewForConfig(cfg)
+		if err != nil {
+			return err
+		}
+
+		// execute the checks
+		if err := engine.ExecuteChecks(); err != nil {
+			return err
+		}
+
+		results := engine.Results()
+
+		// return results to the user and then close output files
+		formattedResults, err := formatter.Format(results)
+		if err != nil {
+			return err
+		}
+
+		fmt.Fprint(os.Stdout, string(formattedResults))
```

Comment:
I know I highlighted this already, but I do see this changed from what I remember seeing before. This still overwrites preflight.log.

Given the following execution:

```
$ PFLT_LOGLEVEL=trace ../preflight check container quay.io/komish/preflight-test-container-passes:latest
time="2021-07-28T11:11:49-05:00" level=info msg="config file not found, proceeding without it"
time="2021-07-28T11:11:49-05:00" level=info msg="target image: quay.io/komish/preflight-test-container-passes:latest"
time="2021-07-28T11:11:49-05:00" level=info msg="downloading image"
time="2021-07-28T11:11:51-05:00" level=info msg="running check: HasRequiredLabel"
time="2021-07-28T11:11:51-05:00" level=info msg="check completed: HasRequiredLabel" result=true
time="2021-07-28T11:11:51-05:00" level=info msg="running check: BasedOnUbi"
time="2021-07-28T11:11:53-05:00" level=info msg="check completed: BasedOnUbi" result=true
time="2021-07-28T11:11:53-05:00" level=info msg="running check: HasLicense"
time="2021-07-28T11:11:58-05:00" level=info msg="1 Licenses found"
time="2021-07-28T11:11:58-05:00" level=info msg="check completed: HasLicense" result=true
time="2021-07-28T11:11:58-05:00" level=info msg="running check: HasMinimalVulnerabilities"
time="2021-07-28T11:11:58-05:00" level=debug msg="Oval file dir: oval-790686789"
time="2021-07-28T11:11:58-05:00" level=debug msg="Oval file path: oval-790686789/rhel-8.oval.xml.bz2"
time="2021-07-28T11:11:58-05:00" level=error msg="unable to execute oscap-podman on the image: This script cannot run in rootless mode.\nvuln.html"
time="2021-07-28T11:11:58-05:00" level=error msg="unable to scan the image: exit status 1"
time="2021-07-28T11:11:58-05:00" level=debug msg="Stdout: []"
time="2021-07-28T11:11:58-05:00" level=info msg="check completed: HasMinimalVulnerabilities" ERROR="failed to scan image: exit status 1" result="failed to scan image: exit status 1"
time="2021-07-28T11:11:58-05:00" level=info msg="running check: HasUniqueTag"
time="2021-07-28T11:11:58-05:00" level=trace msg="running skopeo with the following invocation[list-tags docker://quay.io/komish/preflight-test-container-passes]"
time="2021-07-28T11:11:59-05:00" level=debug msg="detected these tags for quay.io/komish/preflight-test-container-passes: [latest 0.0.1]"
time="2021-07-28T11:11:59-05:00" level=info msg="check completed: HasUniqueTag" result=true
time="2021-07-28T11:11:59-05:00" level=info msg="running check: HasNoProhibitedPackages"
time="2021-07-28T11:12:01-05:00" level=info msg="check completed: HasNoProhibitedPackages" result=true
time="2021-07-28T11:12:01-05:00" level=info msg="running check: RunAsNonRoot"
time="2021-07-28T11:12:03-05:00" level=debug msg="the runtime user id is 1000"
time="2021-07-28T11:12:03-05:00" level=info msg="check completed: RunAsNonRoot" result=true
time="2021-07-28T11:12:03-05:00" level=info msg="running check: LayerCountAcceptable"
time="2021-07-28T11:12:03-05:00" level=debug msg="detected 4 layers in image"
time="2021-07-28T11:12:03-05:00" level=info msg="check completed: LayerCountAcceptable" result=true
{
    "image": "quay.io/komish/preflight-test-container-passes:latest",
    "Status": "FAILED",
    "validation_lib_version": {
        "version": "0.0.0",
        "commit": "6a65a8a7cd3d558c558d944702a72efe5ef89ec6"
    },
    "results": {
        ... truncated for brevity ...
    }
}
```

My logfile only contained a partial result:

```
$ cat preflight.log 
time="2021-07-28T11:12:00-05:00" level=warning msg="The number of prohibited package found in the container image: 0"
time="2021-07-28T11:12:01-05:00" level=info msg="check completed: HasNoProhibitedPackages" result=true
time="2021-07-28T11:12:01-05:00" level=info msg="running check: RunAsNonRoot"
time="2021-07-28T11:12:03-05:00" level=debug msg="the runtime user id is 1000"
time="2021-07-28T11:12:03-05:00" level=info msg="check completed: RunAsNonRoot" result=true
time="2021-07-28T11:12:03-05:00" level=info msg="running check: LayerCountAcceptable"
time="2021-07-28T11:12:03-05:00" level=debug msg="detected 4 layers in image"
time="2021-07-28T11:12:03-05:00" level=info msg="check completed: LayerCountAcceptable" result=true
```

How much the log file contains is a function of the location of the check that utilizes `check run`. Our checks are run in a random order every execution, so if a check that utilizes `check run` is executed in the beginning, we might see more logs. If it's executed at the end, we see less logs.

I think you're running into this because we initialize our logger in the `init()` function for package `cmd`, and we, by default, write to stdout and preflight.log if the user has not specified a configuration.

https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/main/cmd/root.go#L28

You'll probably need to change the built-in logger's outputs to work around this.